qiskit/backends/_projectq_simulator.py

```diff
@@ -0,0 +1,256 @@
+
+    def run_circuit(self, circuit):
+        """Run a circuit and return a single Result.
+
+        Args:
+            circuit (dict): JSON circuit from qobj circuits list
+
+        Returns:
+            A dictionary of results which looks something like::
+
+                {
+                "data":
+                    {  #### DATA CAN BE A DIFFERENT DICTIONARY FOR EACH BACKEND ####
+                    "counts": {’00000’: XXXX, ’00001’: XXXXX},
+                    "time"  : xx.xxxxxxxx
+                    },
+                "status": --status (string)--
+                }
+        """
+        ccircuit = circuit['compiled_circuit']
+        self._number_of_qubits = ccircuit['header']['number_of_qubits']
+        self._number_of_clbits = ccircuit['header']['number_of_clbits']
+        self._quantum_state = 0
+        self._classical_state = 0
+        cl_reg_index = [] # starting bit index of classical register
+        cl_reg_nbits = [] # number of bits in classical register
+        clbit_index = 0
+        qobj_quregs = OrderedDict(_get_register_specs(
+            ccircuit['header']['qubit_labels']))
+        eng = MainEngine(backend=self._sim)
+        for cl_reg in ccircuit['header']['clbit_labels']:
+            cl_reg_nbits.append(cl_reg[1])
+            cl_reg_index.append(clbit_index)
+            clbit_index += cl_reg[1]
+        # let circuit seed override qobj default
+        if 'config' in circuit:
+            if 'seed' in circuit['config']:
+                if circuit['config']['seed'] is not None:
+                    self._sim._simulator = CppSim(circuit['config']['seed'])
+        outcomes = []
+        for shot in range(self._shots):
+            self._quantum_state = np.zeros(1 << self._number_of_qubits,
+                                           dtype=complex)
+            self._quantum_state[0] = 1
+            # initialize starting state
+            self._classical_state = 0
+            unmeasured_qubits = list(range(self._number_of_qubits))
+            projq_qureg_dict = OrderedDict(((key,eng.allocate_qureg(size))
+                                            for key, size in
+                                            qobj_quregs.items()))
+            qureg = [qubit for sublist in projq_qureg_dict.values()
+                     for qubit in sublist]
+            # Do each operation in this shot
+            for operation in ccircuit['operations']:
+                if 'conditional' in operation:
+                    mask = int(operation['conditional']['mask'], 16)
+                    if mask > 0:
+                        value = self._classical_state & mask
+                        while (mask & 0x1) == 0:
+                            mask >>= 1
+                            value >>= 1
+                        if value != int(operation['conditional']['val'], 16):
+                            continue
+                # Check if single  gate
+                if operation['name'] in ['U', 'u3']:
```

Comment:
> is there a way to set a custom single qubit matrix?
@thomashaener, @damiansteiger Please feel free to correct.

Sorry, I didn't see that question earlier.
Yes, it is possible to define gates with custom matrix in which case one can implement u3 directly instead of using Rz Ry Rz. This will speed up the simulation as well.

Should you decide to circumvent the MainEngine (fastest option) as suggested in the previous comment, then you can directly provide a list of lists which specifies the gate matrix instead of using a ProjectQ BasicGate object (`X.matrix.tolist()`)

For a X, it would be:
```
self._sim.apply_controlled_gate([[0,1],[1,0]], qubit_id, [])
```
Just provide the matrix of u3 depending on params

Should you decide to keep the MainEngine, then you could either create a full u3 gate class
```
from projectq.ops import BasicGate
def u3_gate(BasicGate):
    def __init__(self, params):
        BasicGate.__init__(self)
        self.params = params
    @property
    def matrix(self):
        return np.matrix([[0, 1], [1, 0]]) #enter correct matrix depending on self.params
...
u3_gate(params) | qubit

```
or the more on the fly option:
```
u3_gate = BasicGate()
u3.matrix = np.matrix([[0, 1], [1, 0]]) #enter correct matrix depending on params
...
u3_gate | qubit
```